### PR TITLE
File CDK: Add analytics messages for parser usage

### PR DESCRIPTION
**Pull request by flash1293**
## What

Related to https://github.com/airbytehq/airbyte/issues/31497

This PR adds some analytics messages to be able to track how often certain file format parsers are used in practice. This is building on the new analytics handling specified here: https://docs.google.com/document/d/1C6RXbVKtQNZwj3Vey1JbX_FImZrAGyaSUdBY75rZPk0/edit#heading=h.fa72b39y2m99

## How
* Upgrade protocol models to get analytics models (https://github.com/airbytehq/airbyte-protocol/pull/45)
* Add helper to build analytics message
* Define the read method in the file based source to emit analytics messages for the number of streams using a certain parser implementation
* Extend the scenario tester with validating analytics messages
* Add a scenario to test analytics messages that should be emitted

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py [hunk 1]
```diff
@@ -111,6 +113,23 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
...
+    ) -> Iterator[AirbyteMessage]:
+        yield from super().read(logger, config, catalog, state)
+        # count streams using a certain parser
+        stream_by_parser: Dict[str, int] = defaultdict(lambda: 0)
+        parsed_config = self._get_parsed_config(config)
+        for stream in parsed_config.streams:
+            stream_by_parser[stream.format.filetype] += 1
+
+        for (parser, count) in stream_by_parser.items():
+            yield create_analytics_message(f"file-cdk-{parser}-stream-count", count)
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py:131]:
```suggestion
        parsed_config = self._get_parsed_config(config)
        for parser, count in Counter(stream.format.filetype for stream in parsed_config.streams).items():
            yield create_analytics_message(f"file-cdk-{parser}-stream-count", count)
```
(`Counter` is imported from the standard library's `collections` module.)

**Review by clnoll — APPROVED** [review]
Looks great @flash1293! Just one small nit.

**Review by aaronsteers — APPROVED** [review]
LGTM! 

Should we add something to the CDK docs (or other docs) explaining the expected usage pattern?

File: airbyte-cdk/python/airbyte_cdk/utils/analytics_message.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+from airbyte_cdk.models import AirbyteAnalyticsTraceMessage, AirbyteMessage, AirbyteTraceMessage, TraceType, Type
+
+
+def create_analytics_message(type: str, value: Optional[Any]) -> AirbyteMessage:
+    return AirbyteMessage(
+        type=Type.TRACE,
+        trace=AirbyteTraceMessage(
+            type=TraceType.ANALYTICS,
+            emitted_at=time.time() * 1000,
+            analytics=AirbyteAnalyticsTraceMessage(type=type, value=str(value) if value is not None else None),
```
[the remark this answers is not in the document]
Comment (flash1293) [reply airbyte-cdk/python/airbyte_cdk/utils/analytics_message.py:13]: yeah, it's brand new, this is the first usage :)

[the remark this answers is not in the document]
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/utils/analytics_message.py:13]: Nice!

**Comment by flash1293** [comment]
> Should we add something to the CDK docs (or other docs) explaining the expected usage pattern?

That's a good point, I will prepare a notion page for that and do a quick presentation in the dev meeting to make people aware

**Pull request merged**